ProjFS.Mac/PrjFSKext/PrjFSKext/AssociativeArray.hpp

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include "Locks.hpp"
+#include "Memory.hpp"
+#include "KextCppUtilities.hpp"
+
+struct ResizeResult
+{
+    bool   memoryWasUsed;
+    uint32_t releasedMemorySize;
+    void*  releasedMemory;
+};
+
+template <typename KEY, typename VALUE, bool EQUAL_FN(const KEY&, const VALUE&)>
+class AssociativeArray
```

Comment:
@pmj thanks again for the details and for bearing with me as I continue to learn the details of our kext and kext programming in general.

Now that I have a better understanding of the fix for the underlying issue (5177369) and the associative array implementation, I wanted to get your thoughts on a few approaches for simplifying the changes:

_Make the array a fixed size_ 
 
The number of `mount_t` mount points will always be <= the number of virtualization roots on disk, and we don't anticipate users having a lot (e.g. 30+) of virtualization roots.  If we picked a reasonable fixed size (128?) we could avoid the complexity (and failure modes) of resizing the array.

If there were more `mount_t`s than the kext could store it could always fall back to the current behavior of not resetting the TTL and log a warning.

**NOTE:** This suggestion assumes that virtualization roots created in the same partition would share a common `mount_t`, and so our functional tests creating a bunch of test roots in the same folder (`/GVFS.FT`) would not result in a bunch of different `mount_t` entries in the array. 

_Make the array specific to `UsedMountPoint`_

It looks like there's a fair bit of complexity in the implementation of `AssociativeArray` that comes from making it generic.  Did you have other uses in mind for `AssociativeArray`?  If we don't have any other uses for it in the near-term it could be renamed to `MountPointArray` and simplified so that it only works with `UsedMountPoint`s.

(In another comment I saw you mentioned wanting to improve our handling of the virtualization root array, but it wasn't clear to me if you had `AssociativeArray` in mind for that task.)

Reply:
I'll address the suggestions in reverse order as the reasoning in my response makes more sense that way:

> _Make the array a fixed size_
> 
> The number of `mount_t` mount points will always be <= the number of virtualization roots on disk,

It's actually the number of virtualisation roots with active providers, which is <= the number on disk; and the number of mount points will usually be much smaller still.

> and we don't anticipate users having a lot (e.g. 30+) of virtualization roots. If we picked a reasonable fixed size (128?) we could avoid the complexity (and failure modes) of resizing the array.

It's definitely an option, although I feel the complexity is now fairly well contained. We initially had a fixed size array for the virtualisation roots which caused us a lot of problems in the functional tests; of course, for that data structure, the failure to grow is much more serious. As you say, we can just not reset the TTL, which is in fact what this PR does if for some reason the resize fails altogether.

My second argument against this is also my response to the other suggestion:

Reply:
> _Make the array specific to `UsedMountPoint`_
> 
> It looks like there's a fair bit of complexity in the implementation of `AssociativeArray` that comes from making it generic.

Looking over the code, I actually can't find that many instances of code that's not necessary for `UsedMountPoint` - that struct doesn't have constructors & destructors, so invoking them could be removed, which simplifies the resizing code ever so slightly. I'm counting about 9 LOC from about 200, so 5%. I'm not sure that's worth the saving? Granted, we can also remove the move semantics helper code (KextCppUtilities.hpp) but I can remove that anyway if preferred, while still keeping the data structure generic.

> Did you have other uses in mind for `AssociativeArray`? If we don't have any other uses for it in the near-term it could be renamed to `MountPointArray` and simplified so that it only works with `UsedMountPoint`s.
> 
> (In another comment I saw you mentioned wanting to improve our handling of the virtualization root array, but it wasn't clear to me if you had `AssociativeArray` in mind for that task.)

Right, that one isn't really an associative array - the virtualisation root handles are just indices, and while we've got some linear searches that really should be replaced with something more efficient (`FindRootAtVnode_Locked()` & `FindUnusedIndex_Locked()`) those are not the primary lookup means; the way to speed them up would be via an ancillary data structure. (e.g. vnode -> root handle hash table for active providers, fsid/inode -> root handle for offline roots)

However, taking on your earlier idea of passing a lambda to the `Find` function, we could isolate the resizing and linear search logic in a simple generic `Vector` class; the associative array could be based on top of this, while the virtualisation root array uses the vector directly. Given that we definitely want the virtualisation root array to be resizable, this would concentrate the resizing logic for both arrays into one generic implementation, with the instantiation for the roots array getting more testing, but as it's the same code, we can be more confident about resizes of the mount point array too.

Reply:
> All of this might be moot, however, if we can avoid the complexity of having to drop\reacquire the lock completely (see other comment just posted to ProjFS.Mac/PrjFSKext/PrjFSKext/AssociativeArray.hpp).

The new lock won't help with reallocating the virtualisation root array though.